pkg/api/customization/cluster/actionhandler.go

```diff
@@ -273,43 +285,95 @@ func (a ActionHandler) ExportYamlHandler(actionName string, action *types.Action
+	app, err := monitoring.GetApp(a.AppClient, monitoring.OwnedAppListOptions(appName, appTargetNamespace))
 	if err != nil {
-		rtn["message"] = "none existent Cluster"
-		apiContext.WriteResponse(http.StatusBadRequest, rtn)
+		return httperror.WrapAPIError(err, httperror.ServerError, "none Cluster Monitoring app deploying")
+	}
+	if app == nil {
+		return httperror.NewAPIError(httperror.ServerError, "none Cluster Monitoring app deploying")
+	}
 
-		return errors.Wrapf(err, "failed to get Cluster by ID %s", apiContext.ID)
+	// TODO, need to support `map[string]string` as entry value type in norman Builder.convertMap
+	answers, err := convert.EncodeToMap(app.Spec.Answers)
+	if err != nil {
+		return httperror.WrapAPIError(err, httperror.ServerError, "failed to parse response")
+	}
+	apiContext.WriteResponse(http.StatusOK, map[string]interface{}{
+		"answers": answers,
+		"type":    "monitoringOutput",
+	})
+	return nil
+}
+
+func (a ActionHandler) editMonitoring(actionName string, action *types.Action, apiContext *types.APIContext) error {
+	cluster, err := a.ClusterClient.Get(apiContext.ID, metav1.GetOptions{})
+	if err != nil {
+		return httperror.WrapAPIError(err, httperror.NotFound, "none existent Cluster")
 	}
 	if cluster.DeletionTimestamp != nil {
-		rtn["message"] = "deleting Cluster"
-		apiContext.WriteResponse(http.StatusBadRequest, rtn)
+		return httperror.NewAPIError(httperror.InvalidType, "deleting Cluster")
+	}
 
-		return fmt.Errorf("unable to operate deleting %s Cluster", apiContext.ID)
+	if !cluster.Spec.EnableClusterMonitoring {
+		return httperror.NewAPIError(httperror.InvalidState, "disabling Monitoring")
+	}
+
+	data, err := ioutil.ReadAll(apiContext.Request.Body)
+	if err != nil {
+		return httperror.WrapAPIError(err, httperror.InvalidBodyContent, "unable to read request content")
+	}
+	var input v3.MonitoringInput
+	if err = json.Unmarshal(data, &input); err != nil {
+		return httperror.WrapAPIError(err, httperror.InvalidBodyContent, "failed to parse request content")
+	}
+
+	cluster = cluster.DeepCopy()
+	cluster.Annotations = monitoring.AppendAppOverwritingAnswers(cluster.Annotations, string(data))
+
+	_, err = a.ClusterClient.Update(cluster)
+	if err != nil {
+		return httperror.WrapAPIError(err, httperror.ServerError, "failed to upgrade monitoring")
+	}
+
+	apiContext.WriteResponse(http.StatusOK, map[string]interface{}{})
+	return nil
+}
+
+func (a ActionHandler) enableMonitoring(actionName string, action *types.Action, apiContext *types.APIContext) error {
+	cluster, err := a.ClusterClient.Get(apiContext.ID, metav1.GetOptions{})
+	if err != nil {
+		return httperror.WrapAPIError(err, httperror.NotFound, "none existent Cluster")
+	}
+	if cluster.DeletionTimestamp != nil {
+		return httperror.NewAPIError(httperror.InvalidType, "deleting Cluster")
 	}
 
 	if cluster.Spec.EnableClusterMonitoring {
-		apiContext.WriteResponse(http.StatusOK, rtn)
+		apiContext.WriteResponse(http.StatusOK, map[string]interface{}{})
```

Comment:
If you don't return anything here, you would like to use `http.StatusNoContent` instead of `http.StatusNoContent`.

Reply:
Refactored